airflow/models/dagrun.py

```diff
@@ -580,11 +591,11 @@ def recalculate(self) -> _UnfinishedStates:
             self.set_state(DagRunState.FAILED)
             if execute_callbacks:
                 dag.handle_callback(self, success=False, reason='task_failure', session=session)
-            elif dag.has_on_failure_callback:
+            elif dag.has_on_failure_callback or notification:
                 from airflow.models.dag import DagModel
 
                 dag_model = DagModel.get_dagmodel(dag.dag_id, session)
-                callback = DagCallbackRequest(
+                cb = DagCallbackRequest(
```

Comment:
`cb` vs `callback` doesn't really signal what's the difference here. perhaps you can find a way to signal the intention.  e.g. `returned_callback` vs `callback`

something like that

Reply:
Changed that, also the notification now does not rely on `DagCallbackRequest` - since we don't always have it, like in `BackfillJob`.